src/backend/distributed/operations/shard_cleaner.c

```diff
@@ -132,8 +190,77 @@ TryDropOrphanedShards(bool waitForLocks)
 }
 
 
+static int
+DropOrphanedShardsForCleanup()
+{
+	/* Only runs on Coordinator */
+	if (!IsCoordinator())
+	{
+		return 0;
+	}
+
+	List *cleanupRecordList = ListCleanupRecords();
+
+	int removedShardCountForCleanup = 0;
+	int failedShardCountForCleanup = 0;
+	CleanupRecord *record = NULL;
+
+	foreach_ptr(record, cleanupRecordList)
+	{
+		/* We only support one resource type at the moment */
+		if (record->objectType != CLEANUP_SHARD_PLACEMENT)
+		{
+			ereport(WARNING, (errmsg("Invalid object type %d for cleanup record ",
+									 record->objectType)));
+			continue;
+		}
+
+		/* Advisory locks are reentrant */
+		if (!TryLockOperationId(record->operationId))
+		{
+			/* operation that the cleanup record is part of is still running */
+			continue;
+		}
+
```

Comment:
should we let go of the advisory lock now? no one else will take it, but there's a limit to how many advisory locks we can reasonably take

Reply:
Now that I am thinking about it more, are you proposing to release a lock if operationId changes when iterating on the loop? That would be hard to know for sure as there is no sort order.

Reply:
It's mainly odd to me that we get arbitrarily many operation ID locks here. On the other hand, that is probably the safe thing to do in case we ever reuse operation IDs.

Reply:
There is an API 'LockHeldByMe' : https://github.com/postgres/postgres/blob/8ec569479fc28ddd634a13dc100b36352ec3a3c2/src/include/storage/lock.h#L563
we could check this API and take a lock only if it returns false.